src/core/server/saved_objects/validation/types.ts

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { ObjectType } from '@kbn/config-schema';
+
+/**
+ * The custom validation function if @kbn/config-schema is not a valid solution for your specific plugin requirements.
+ *
+ * Be careful not to mutate the provided attributes.
+ *
+ * @example
+ * The validation should look something like:
+ * ```typescript
+ * const myAttributesValidation: SavedObjectsValidationFunction = ({ attributes }) => {
+ *   if (typeof attributes.bar !== 'string') {
+ *     throw new Error(`[bar]: expected value of type [string] but got [${typeof attributes.bar}]`);
+ *   }
+ * }
+ * ```
+ *
+ * @public
+ */
+export type SavedObjectsValidationFunction = (data: { attributes: unknown }) => void;
+
+/**
+ * Allowed property validation options: either @kbn/config-schema validations or custom validation functions.
+ *
+ * See {@link SavedObjectsValidationFunction} for custom validation.
+ *
+ * @public
+ */
+export type SavedObjectsValidationSpec = ObjectType | SavedObjectsValidationFunction;
```

Comment:
Why do we want to support custom validators? I'd start with supporting `@kbn/config-schema` only and relax requirements only (if) it's needed.

Reply:
Initially the idea was to keep the DX similar to our http route validations since most devs are already familiar with those, and downstream teams requesting custom validation functions felt like an inevitability.

However, you raise a fair point: as this function is synchronous & we've decided to only expose the `attributes` atm, most use cases I can think of for this could already be achieved using the custom validators already provided by config-schema. One such use case would be stringified json such as is used for `visState`, where plugins may want to parse the json and validate fields individually.

With that in mind, I'm okay removing the added complexity for the time being & waiting to see if teams end up needing anything more than what config-schema already provides out of the box.

Reply:
@pgayvallet WDYT, since we had already discussed the custom validators?

Reply:
In theory, anything can be done with custom `validate` function of our config schema (even things like parsing json to validate the content), so I don't have objections starting with only config schema validation and waiting to see if teams eventually need more than that

Reply:
okay, I went ahead and removed it for now to simplify